src/Type/Union/MenuItemObjectUnionType.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * MenuItemObjectUnionType
+ */
+
+namespace WPGraphQL\Type\Union;
+
+use GraphQL\Type\Definition\UnionType;
+use WPGraphQL\Types;
+
+/**
+ * Class MenuItemObjectUnionType
+ *
+ * Navigation menus comprise menu items that reference an object, which can be
+ * a post object, a taxonomy term object, or a custom link.
+ */
+class MenuItemObjectUnionType extends UnionType {
+	/**
+	 * An array of the possible types that can be resolved by this union.
+	 *
+	 * @var array
+	 */
+	private static $possible_types;
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$config = [
+			'name'  => 'MenuItemObjectUnion',
+			'types' => self::get_possible_types(),
+			'resolveType' => function( $object ) {
+				// Custom link / menu item
+				if ( $object instanceof \WP_Post && 'nav_menu_item' === $object->post_type ) {
+					return Types::menu_item();
+				}
+
+				// Post object
+				if ( $object instanceof \WP_Post && ! empty( $object->post_type ) ) {
+					return Types::post_object( $object->post_type );
+				}
+
+				// Taxonomy term
+				if ( $object instanceof \WP_Term && ! empty( $object->taxonomy ) ) {
+					return Types::term_object( $object->taxonomy );
+				}
+
+				return null;
+			},
+		];
+
+		parent::__construct( $config );
+	}
+
+	/**
+	 * This defines the possible types that can be resolved by this union
+	 *
+	 * @return array An array of possible types that can be resolved by the union
+	 * @since 0.0.5
+	 */
+	public static function get_possible_types() {
+		if ( is_array( self::$possible_types ) ) {
+			return self::$possible_types;
+		}
+
+		// We could restrict this further using `show_in_nav_menus`, but it's of
+		// questionable utility since we'd only be creating more work for those
+		// that want to implement custom resolution of menu items.
```

Comment:
I think I’m ok with just restricting to `show_in_graphql` As you have it. Thanks for calling this out in the comment though. 😀